File: problem_062.c
// -*- compile-command: "gcc -o problem_062 problem_062.c -Wall -lm" -*-
// Copyright (c) 2014 Michael Caldwell
#include <stdio.h>
#include <math.h>
#include <stdbool.h>

unsigned numSignature(unsigned long long number)
{
    unsigned out = 0;
    while (number > 0)
    {
        out += 1 * pow(10, number % 10);
        number /= 10;
    }
    return out;
}

int main(){
    printf("Project Euler - Problem 62:\n"
           "Find the smallest cube for which exactly five permutations of its digits are cube.\n\n");

    unsigned results[10000] = {};

    // Get Signatures for powers of 3
    int i = 0;
    for (i=2 ; i<10000 ; i++)
        results[i] = numSignature(pow(i,3));

/*
    // DEBUG - Print Table
    for (i=2 ; i<10000 ; i++)
        printf("%llu\n", results[i]);
*/

    // Search Results for signatures that are same
    for (i=0 ; i<10000 ; i++)
    {
        int n, count=0;
        for (n=0 ; n<10000 ; n++)
            if (results[n] == results[i])
                count++;
        if (count == 5)
            break;
    }

    printf("%d^3 = %lu\n", i, (unsigned long)pow(i, 3));
    return 0;
}

File: problem_062.swift
#!/usr/bin/xcrun swift
// -*- compile-command: "xcrun -sdk macosx swiftc problem_062.swift -o problem_062_swift" -*-
// Copyright (c) 2016 Michael Caldwell
import Foundation

// Deconstruct the number to make a number signature
func numSig(n: Int) -> Int {
    var n = n
    var sig = 0
    while n > 0 {
        sig += Int(pow(10.0, Float(n % 10)))
        n /= 10
    }
    return sig
}

print("Project Euler - Problem 62:")
print("Find the smallest cube for which exactly five permutations of its digits are cube.\n")

let debug = false

// Generate our gonal numbers in the 4 digit range
var cubes = [Int]()
var dict = [Int:[Int]]()
var result = 0

for i in 0...10000 {
    let cube = i * i * i
    let sig = numSig(n: cube)

    // Check if dict contains signature yet
    if let value = dict[sig] {
        // Check if we have already encountered 4 times
        if value[1] == 4 {
            // We found our answer!
            result = value[0]
            break
        }

        // Increment our count
        dict[sig] = [value[0], value[1] + 1]
    } else {
        // If first time encountering sig, record it
        dict[sig] = [i, 1]
    }

    if debug == true {
        print(i, cube, sig)
    }
}

print(result)
